ts/session/sending/MessageQueue.ts

```diff
@@ -1,54 +1,190 @@
+import { ConversationController } from '../../window';
+
 import { EventEmitter } from 'events';
 import {
   MessageQueueInterface,
   MessageQueueInterfaceEvents,
 } from './MessageQueueInterface';
-import { ContentMessage, OpenGroupMessage } from '../messages/outgoing';
+import {
+  ClosedGroupMessage,
+  ContentMessage,
+  OpenGroupMessage,
+  SessionRequestMessage,
+} from '../messages/outgoing';
 import { PendingMessageCache } from './PendingMessageCache';
-import { JobQueue, TypedEventEmitter } from '../utils';
+import {
+  GroupUtils,
+  JobQueue,
+  SyncMessageUtils,
+  TypedEventEmitter,
+} from '../utils';
+import { PubKey } from '../types';
+import { MessageSender } from '.';
+import { SessionProtocol } from '../protocols';
+import { UserUtil } from '../../util';
 
 export class MessageQueue implements MessageQueueInterface {
   public readonly events: TypedEventEmitter<MessageQueueInterfaceEvents>;
-  private readonly jobQueues: Map<string, JobQueue> = new Map();
-  private readonly cache: PendingMessageCache;
+  private readonly jobQueues: Map<PubKey, JobQueue> = new Map();
+  private readonly pendingMessageCache: PendingMessageCache;
 
   constructor() {
     this.events = new EventEmitter();
-    this.cache = new PendingMessageCache();
-    this.processAllPending();
+    this.pendingMessageCache = new PendingMessageCache();
+    void this.processAllPending();
   }
 
-  public sendUsingMultiDevice(user: string, message: ContentMessage) {
-    throw new Error('Method not implemented.');
+  public async sendUsingMultiDevice(user: PubKey, message: ContentMessage) {
+    const userLinked = await getPairedDevicesFor(user.key);
+    const userDevices = userLinked.map(d => new PubKey(d));
+
+    await this.sendMessageToDevices(userDevices, message);
   }
-  public send(device: string, message: ContentMessage) {
-    throw new Error('Method not implemented.');
+
+  public async send(device: PubKey, message: ContentMessage) {
+    await this.sendMessageToDevices([device], message);
   }
-  public sendToGroup(message: ContentMessage | OpenGroupMessage) {
-    throw new Error('Method not implemented.');
+
+  public async sendMessageToDevices(
+    devices: Array<PubKey>,
+    message: ContentMessage
+  ) {
+    let currentDevices = [...devices];
+
+    // Sync to our devices if syncable
+    if (SyncMessageUtils.canSync(message)) {
+      const currentDevice = await UserUtil.getCurrentDevicePubKey();
+
+      if (currentDevice) {
+        const otherDevices = await getPairedDevicesFor(currentDevice);
+
+        const ourDevices = [currentDevice, ...otherDevices].map(
+          device => new PubKey(device)
+        );
+        await this.sendSyncMessage(message, ourDevices);
+
+        // Remove our devices from currentDevices
+        const ourDeviceContacts = ourDevices.map(device =>
+          ConversationController.get(device.key)
+        );
+        currentDevices = _.xor(currentDevices, ourDeviceContacts);
```

Comment:
```suggestion
        currentDevices = currentDevices.filter(device => ourDevices.some(d => d.key === device.key))
```
This might be easier than mapping it using `ConversationController.get`